cores/esp32/chip-debug-report.cpp

```diff
@@ -0,0 +1,338 @@
+#include "esp_heap_caps.h"
+#include "esp_chip_info.h"
+#include "esp_idf_version.h"
+#include "esp_arduino_version.h"
+#include "esp_rom_spiflash.h"
+#include "esp_flash.h"
+#include "esp_partition.h"
+#include "esp_app_format.h"
+#include "soc/efuse_reg.h"
+#include "soc/rtc.h"
+#include "soc/spi_reg.h"
+#if CONFIG_IDF_TARGET_ESP32S2
+#include "esp32s2/rom/spi_flash.h"
+#endif
+#include "esp_bit_defs.h"
+
+#include "esp32-hal.h"
+#include "esp32-hal-periman.h"
+
+#define printMemCapsInfo(caps) _printMemCapsInfo(MALLOC_CAP_##caps, #caps)
+#define b2kb(b) ((float)b/1024.0)
+#define b2mb(b) ((float)b/(1024.0*1024.0))
+static void _printMemCapsInfo(uint32_t caps, const char * caps_str){
+  multi_heap_info_t info;
+  size_t total = heap_caps_get_total_size(caps);
+  heap_caps_get_info(&info, caps);
+  log_printf("%s Memory Info:\n", caps_str);
+  log_printf("------------------------------------------\n");
+  log_printf("  Total Size        : %8u B (%6.1f KB)\n", total, b2kb(total));
+  log_printf("  Free Bytes        : %8u B (%6.1f KB)\n", info.total_free_bytes, b2kb(info.total_free_bytes));
+  log_printf("  Allocated Bytes   : %8u B (%6.1f KB)\n", info.total_allocated_bytes, b2kb(info.total_allocated_bytes));
+  log_printf("  Minimum Free Bytes: %8u B (%6.1f KB)\n", info.minimum_free_bytes, b2kb(info.minimum_free_bytes));
+  log_printf("  Largest Free Block: %8u B (%6.1f KB)\n", info.largest_free_block, b2kb(info.largest_free_block));
+}
+
+static void printPkgVersion(void){
+  log_printf("  Package           : ");
+#if CONFIG_IDF_TARGET_ESP32
+  uint32_t pkg_ver = REG_GET_FIELD(EFUSE_BLK0_RDATA3_REG, EFUSE_RD_CHIP_PACKAGE);
+  switch(pkg_ver){
+    case EFUSE_RD_CHIP_VER_PKG_ESP32D0WDR2V3: log_printf("D0WD-R2-V3"); break;
+    case EFUSE_RD_CHIP_VER_PKG_ESP32D0WDQ6  : log_printf("D0WD-Q6"); break;
+    case EFUSE_RD_CHIP_VER_PKG_ESP32D0WDQ5  : log_printf("D0WD-Q5"); break;
+    case EFUSE_RD_CHIP_VER_PKG_ESP32D2WDQ5  : log_printf("D2WD-Q5"); break;
+    case EFUSE_RD_CHIP_VER_PKG_ESP32U4WDH   : log_printf("U4WD-H"); break;
+    case EFUSE_RD_CHIP_VER_PKG_ESP32PICOD4  : log_printf("PICO-D4"); break;
+    case EFUSE_RD_CHIP_VER_PKG_ESP32PICOV302: log_printf("PICO-V3-02"); break;
+  }
+#elif CONFIG_IDF_TARGET_ESP32S2
+  uint32_t pkg_ver = REG_GET_FIELD(EFUSE_RD_MAC_SPI_SYS_3_REG, EFUSE_PKG_VERSION);
+  switch (pkg_ver) {
+    case 1: log_printf("FH16"); break;
+    case 2: log_printf("FH32"); break;
+    default: log_printf("%lu", pkg_ver); break;
+  }
+#elif CONFIG_IDF_TARGET_ESP32S3 || CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32C6
+  uint32_t pkg_ver = REG_GET_FIELD(EFUSE_RD_MAC_SPI_SYS_3_REG, EFUSE_PKG_VERSION);
```

Comment:
Is there any documentation on what this pkg_ver is telling us for the S3/C3/etc?
I couldn't find anything yesterday.

Reply:
no and they all return 0 so far. It will probably make more sense in the future. Chips are too new

Reply:
Well the S3 already has seen [quite a few iterations and versions](https://espeasy.readthedocs.io/en/latest/ESPEasy/ESPchips.html#quad-octal-spi-mode).
So I find it very strange (frustrating) that there seems to be no way to read from some efuses whether the chip has embedded flash/PSRAM, its size and how it is wired, without actually access either flash or PSRAM and hopefully using the correct nr. of lanes.

Reply:
you must have skipped the top of the log?

Reply:
It does indeed show whether flash or PSRAM is embedded, but you can't simply see whether octal or quad mode is needed for PSRAM.
Thus it seems impossible to detect which build might be needed when initially flashing the board.

I have 2 boards as an example:
- (A) LilyGo T-dongle, which has 16M quad flash, no PSRAM
- (B) S3 "Wemos dev pro" with 16M quad flash, 8MB octal PSRAM

Flashing a build for quad flash, quad PSRAM will work fine on board A, which doesn't have PSRAM , but on  board B with 8M PSRAM you can't detect it. (and it isn't embedded in the chip, thus the "embedded PSRAM" efuse isn't set)

And when flashing a build with quad flash, octal PSRAM to board A which hasn't (such) PSRAM, you end up with a bootloop.

N.B. this board I have with 16M quad flash and 8M PSRAM (not embedded) does have an official Espressif module, so it would have been nice if Espressif did mark with some efuses how to recognize those features before flashing.

This makes it quite a trial-and-error run to properly detect what features are present on a board.

And you might say, well I don't need the PSRAM, so it doesn't matter, but using GPIO33-37 will still be wired to the PSRAM, thus using them for something else might have side-effects.

You can still read the efuses to try figuring out whether these specific pins are 'powered' via either the internal LDO or by the ESP's Vdd. But this is quite complex as you can also have override efuses active to ignore the bootstrap pin GPIO 45(??) which controls this internal voltage and I'm not 100% sure whether reading those states is actually reading an efuse, or the actual state of the power supply for GPIO 33-37 and thus maybe related to having PSRAM support enabled, etc.

TL;DR
I still think it could have been done a lot simpler by Espressif...

Reply:
There is an efuse set for octal PSRAM and you will get an error when booting in wrong mode. This dump is printed AFTER everything is setup and after any error from PSRAM.

Reply:
> There is an efuse set for octal PSRAM

Which one?
That would be of great help to let a user know there might be a better suited build available if it can be detected what mode to use.
Use case: I'm thinking of something very similar to what Tasmota is doing with their "Safeboot" builds.
So if that build could simply detect the presence of PSRAM (or at least the required SPI mode) without crashing, then it would be a lot more user friendly compared to flashing and hoping it won't end up in a bootloop.

Reply:
hmm I am not finding it in the lists... maybe it was for flash only, though I do not find that one either.

ESP32-S3 so far comes with two PSRAM options:
- 2MB QSPI PSRAM
- 8MB OPI PSRAM

So if you check the markings on the module and you see NxxR8, that means you have OPI PSRAM, R2 will be QSPI

Reply:
> hmm I am not finding it in the lists... maybe it was for flash only, though I do not find that one either.

This is what I'm using so far for the OPI flash wired efuse.
Not sure how useful it is, as it will probably not boot when the wrong one was set.
But esptool.py could perhaps at some point do a check for it...

Label in the datasheet is "flash type" if I remember correctly.

```c++
bool getFlashChipOPI_wired() {
  #ifdef ESP32_CLASSIC
    return false;

  #else
  // Source: https://github.com/espressif/esptool/commit/b25606b95920bd06df87aff9202c7a15377d4a30
  const uint32_t data = REG_GET_BIT(EFUSE_RD_REPEAT_DATA3_REG, BIT(9));
  return data != 0;
  #endif
}
```

Reply:
5.1 supports flash auto-detection and that is turned on for ESP32S3. Should boot either way. The same technique is not possible for PSRAM

Reply:
> 5.1 supports flash auto-detection and that is turned on for ESP32S3. Should boot either way.

That's good to know, will at least save a lot of hassle.

I guess there is some other checking being done when accessing the PSRAM as I've also seen some logs showing a bit more info on the detected PSRAM like brand (??).
So would be really nice if there could still be some runtime check to be found for trying PSRAM SPI mode. It would already be useful if it were only for embedded PSRAM as those are more likely to already have an efuse set indicating it has PSRAM.

Reply:
I have some ESP32-S2 unit here, which does have 2M of embedded PSRAM and 4MB embedded flash.
However the chip features flags do not mention embedded flash nor embedded PSRAM.

<html>
<body>
<!--StartFragment-->

ESP Chip ID: | 5609182 (0x5596DE)
-- | --
ESP Chip Frequency: | 240 MHz
ESP Crystal Frequency: | 40 MHz
ESP APB Frequency: | 80 MHz
ESP Chip Model: | ESP32-S2FNR2
ESP Chip Features: | Wi-Fi bgn
ESP Chip Revision: | 0.0
ESP Chip Cores: | 1

<!--EndFragment-->
</body>
</html>

The `ESP Chip Features` string is generated with this code:

```c++
    String features;

    if (getChipFeatures().wifi_bgn) { features += F("Wi-Fi bgn / "); }
    if (getChipFeatures().bluetooth_ble) { features += F("BLE / "); }
    if (getChipFeatures().ieee_802_15_4) { features += F("IEEE 802.15.4 / "); }
    if (getChipFeatures().embeddedFlash) { features += F("Emb. Flash / "); }
    if (getChipFeatures().embeddedPSRAM) { features += F("Emb. PSRAM"); }
    features.trim();

    if (features.endsWith(F("/"))) { features = features.substring(0, features.length() - 1); }
    addHtml(features);
```

```c++
struct esp32_chip_features {
  bool embeddedFlash{};
  bool wifi_bgn{};
  bool bluetooth_ble{};
  bool bluetooth_classic{};
  bool ieee_802_15_4{};
  bool embeddedPSRAM{};
};

esp32_chip_features getChipFeatures() {
  static esp32_chip_features res;
  static bool loaded = false;
  if (!loaded) {
    esp_chip_info_t chip_info;
    esp_chip_info(&chip_info);

    res.embeddedFlash = chip_info.features & CHIP_FEATURE_EMB_FLASH;
    res.wifi_bgn      = chip_info.features & CHIP_FEATURE_WIFI_BGN;
    res.bluetooth_ble = chip_info.features & CHIP_FEATURE_BLE;
    res.bluetooth_classic = chip_info.features & CHIP_FEATURE_BT;
    res.ieee_802_15_4 = chip_info.features & CHIP_FEATURE_IEEE802154;
    res.embeddedPSRAM = chip_info.features & CHIP_FEATURE_EMB_PSRAM;

    loaded = true;
  }
  return res;
}
```

This is one of those purple "Wemos" modules with the bare chip on the PCB. So both flash and PSRAM are for sure embedded in the chip and thus Espressif should have burnt those efuse bits, right?

In other words, those `chip_info.features` may not represent the correct state.

Reply:
> In other words, those chip_info.features may not represent the correct state.

I can not comment on that. I'm just dumping what ESP-IDF provides.

Reply:
> In other words, those `chip_info.features` may not represent the correct state.

In most cases that I've seen where they don't match up they are using a SoC that doesn't have anything embedded and instead uses external flash / psram. I have yet to find any Espressif branded modules that exhibit different behaviors.